text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,376 @@
+
+Two projects can be built in parallel without troubles.
+
+#### With both set
+
+`CARGO_TARGET_DIR` was present long before `CARGO_TARGET_DIRECTORIES`: backward compatibility is important, so the first always trumps the second,
+there is no mixing going on.
+
+#### Absolute and relative paths
+
+`CARGO_TARGET_DIR` can be either a relative or absolute path, which makes sense since it's mostly intended for a single project, which can then
+work from its own position to configure the target directory.
+
+On the other hand `CARGO_TARGET_DIRECTORIES` is intended to be used with several projects, possibly completely unrelated to each other. As such,
+it does not accept relative paths, only absolute ones. If a compelling use case is present for a relative path, it can added in the future as a
+backward-compatible change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Setting `CARGO_TARGET_DIRECTORIES`
+
+The option is similar to `CARGO_TARGET_DIR` and can be set in the same places. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-directories = "/absolute/path/to/target/directories"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIRECTORIES="/absolute/path/to/target/directories" cargo build`
+- Through the command line flag: `cargo build --target-directories /absolute/path/to/target/directories`
+
+The given path must be absolute: setting `CARGO_TARGET_DIRECTORIES` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).
+
+## Resolution order relative to `CARGO_TARGET_DIR`
+
+The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:
+
+`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-directories` > `CARGO_TARGET_DIRECTORIES` > `target-directories = ...`
+
+See the final section for discussion on how to make `CARGO_TARGET_DIRECTORIES` the default.
+
+## Naming
+
+In the example in the previous section, using `CARGO_TARGET_DIRECTORIES` with `cargo build` produces named subdirectories. The name of those is partially deterministic:
+it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.
```

Comment:
What do we see as the likely upper bound on projects in `CARGO_TARGET_DIRECTORIES`?

On at least some platforms, having a directory with a lot of files/directories can be slow and this is worked around by pulling out a level or two of prefixes from the name and making those parent directories.  I'm unsure if a directory name would have enough entropy to do this but say we did, then the path would be `/cargo-caches/wo/rk/-project-<hash>/debug/...`

Reply:
I have encountered that in the past but that was with thousands of directories, I don't know if that'll ever be reached in our case but as always in computing, all limits are reached eventually so a little thinking now could prevent much pain later

I want to discuss it more because I'm not sure my napkin-quick design is any good so I'll add a section in *Unresolved Questions* explaining my ideas and open a review comment there